plugin.go

```diff
@@ -225,11 +234,78 @@ func (s *Plugin) Middleware(next http.Handler) http.Handler { //nolint:gocognit,
 			}
 		}
 
+		if s.gzip && r.Header.Get("Accept-Encoding") == "gzip" {
+			s.log.Debug("gzip compression requested")
+			cf, err := s.compressContent(fp, f)
+			if err != nil {
+				s.log.Error("failed to compress content", zap.Error(err))
+				http.Error(w, "Internal Server Error", http.StatusInternalServerError)
+				return
+			}
+			err = f.Close()
+			if err != nil {
+				s.log.Error("failed to close file", zap.Error(err))
+			}
+			f = cf
+			// No new defer to close needed for cf. f is already closed, previous defer exist
+			w.Header().Set("Content-Encoding", "gzip")
+		}
+
 		// we passed all checks - serve the file
 		http.ServeContent(w, r, finfo.Name(), finfo.ModTime(), f)
 	})
 }
 
+func (s *Plugin) createCompressedFile(originalFile http.File, compressedPath string) error {
+	cacheMutex.Lock()
+	defer cacheMutex.Unlock()
+
+	// We check again to avoid a race between threads.
+	if _, err := os.Stat(compressedPath); err == nil {
+		return nil
+	}
+	compressedFile, err := os.Create(s.cfg.Dir + compressedPath)
+	if err != nil {
+		return err
+	}
+	defer func() {
```

Comment:
Too many unnececarry defers. There're no errors returning by another function calls, to overload a stack with 2 defers in a row.